app/controllers/user/sessions_controller.rb

```diff
@@ -0,0 +1,47 @@
+class User::SessionsController < Devise::SessionsController
+  def create
+    if session.has_key?(:user_sign_in_uid)
+      self.resource = User.find(session[:user_sign_in_uid])
+      session.delete(:user_sign_in_uid)
+    else
+      self.resource = warden.authenticate!(auth_options)
+    end
+
+    if resource.active_for_authentication? && resource.otp_module_enabled?
+      if params[:user][:otp_attempt].blank?
+        session[:user_sign_in_uid] = resource.id
+        sign_out(resource)
+        redirect_to user_two_factor_entry_url
+      else
+        if resource.authenticate_otp(params[:user][:otp_attempt])
+          continue_sign_in(resource, resource_name)
+        else
+          sign_out(resource)
+          flash[:error] = t('devise.failure.invalid')
+          redirect_to root_url
```

Comment:
not sure if redirecting to the root_url is a nice thing to do; maybe a redirect to the sign in path would be better here?